xarray/core/computation.py

```diff
@@ -343,8 +344,10 @@ def apply_dataset_ufunc(func, *args, **kwargs):
     list_of_coords = build_output_coords(args, signature, exclude_dims)
 
     args = [getattr(arg, 'data_vars', arg) for arg in args]
+
     result_vars = apply_dict_of_variables_ufunc(
-        func, *args, signature=signature, join=join, fill_value=fill_value)
+        func, *args, signature=signature, join=data_vars_join,
```

Comment:
What is the expected behavior of the following?
```
ds0 = xr.Dataset({'a': ('x', [1, 2]), 'x': [0, 1]})
ds1 = xr.Dataset({'b': ('x', [99, 3]), 'x': [1, 2]})

def add(a, b, join, data_vars_join, **kwargs):
        return apply_ufunc(operator.add, a, b, join=join,
                           data_vars_join=data_vars_join, **kwargs)

out = add(ds0, ds1, 'outer', 'outer')
```
Currently, it raises a TypeError at `result_vars[name] = func(*variable_args)` of `apply_dict_of_variables_ufunc`.  

On the other hand, if you run
```
out = add(ds0, ds2, 'outer', 'outer', dataset_fill_value=np.nan)
```
you get
```
>> out
<xarray.Dataset>
Dimensions:  (x: 3)
Coordinates:
  * x        (x) int64 0 1 2
Data variables:
    a        (x) float64 nan nan nan
    b        (x) float64 nan nan nan
```
While I can understand why the code renders such result, as a user, I was hoping for something more like
```
<xarray.Dataset>
Dimensions:  (x: 3)
Coordinates:
  * x        (x) int64 0 1 2
Data variables:
    a        (x) float64 1.0 2.0 nan
    b        (x) float64 nan 99.0 3.0
```
which would be a little more useful.

would we want to make changes for this?